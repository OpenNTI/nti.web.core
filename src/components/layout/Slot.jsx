/** @typedef {string|JSX.Element} SlotIdentifier - either a name or React component */
/** @typedef {SlotIdentifier|[SlotIdentifier]} Slot */
/** @typedef {(cmp: JSX.Element) => boolean} Matcher */
/** @typedef {(cmp: JSX.Element, index:number) => JSX.Element} Mapper */
/** @typedef {(slot:Slot, cmp: JSX.Element, index:number) => JSX.Element} ElementMapper */
/** @typedef {(slot: Slot, cmp: JSX.Element, index: number) => boolean | boolean} ElementFilter - if === true filter any child that does not match a slot, if a function filter by the return value*/

import React from 'react';

const Matchers = {
	string:
		slot =>
		({ type, props: { slot: slotProp } }) =>
			slotProp === slot || type.slot === slot,

	function:
		slot =>
		({ type }) =>
			type === slot,

	object:
		slot =>
		({ type }) =>
			type === slot,

	undefined:
		() =>
		({ props: { slot } }) =>
			!slot,
};

function buildMatcher(slot, inverse) {
	const matcher = Matchers[typeof slot]?.(slot);

	if (!matcher) {
		throw new Error('Unknown slot value');
	}

	return inverse
		? (...args) => {
				return !matcher(...args);
		  }
		: matcher;
}

function buildMatchers(slots, inverse) {
	const matchers = slots.map(s => buildMatcher(s, inverse));

	return (...args) => matchers.every(m => m(...args));
}

/**
 * Generate a matcher for the given slot identifiers
 *
 * @private
 * @param {Slot} slot
 * @param {boolean} inverse
 * @returns {Matcher}
 */
export function buildMatchFn(slot, inverse) {
	return Array.isArray(slot)
		? buildMatchers(slot, inverse)
		: buildMatcher(slot, inverse);
}

/**
 * This component provides a mechanism by which a parent component
 * can render provided children into designated 'slots' in its markup
 *
 * Types of Slot identifiers:
 * 1.) String - looks for the child with a 'slot' prop === the slot identifier
 *
 * 		const Parent = (props) => (
 * 			<section>
 * 				<header>
 * 					<Slot slot="header" {...props} /> <== render children with a slot="header" prop here
 * 				</header>
 * 				<Slot {...props} /> <== render children without a slot prop here
 * 			</section>
 * 		)
 *
 * 		<Parent>
 * 			<h1 slot="header">{pageTitle}</h1> <== gets rendered into Parent's <header> element
 * 		</Parent>
 *
 * 2.) Component - looks for a child with a type === the slot identifier
 *
 * 		const SubCmpFooter = () => (<div>Sub Component Footer!</div>);
 * 		const Parent = (props) => (
 * 			<section>
 * 				<Slot exclude={SubCmpFooter} />
 * 				<Slot slot={SubCmpFooter} />
 * 			</section
 * 		);
 *
 * 		<Parent>
 * 			<SubCmpFooter />
 * 			<div>
 * 				Arbitrary Content
 * 			</div>
 * 		</Parent>
 *
 * Matchers:
 * 1.) slot - look for children that match the slot identifer
 * 1a) [slot] - look for children that match one of the slot identifiers
 * 2.) exclude - look for children that do not match the slot identifier
 * 2a.) [exclude] - look for children that do not match any of the slot identifiers
 *
 * @param {{slot:Slot, exclude: Slot, map: Mapper}} props
 * @returns {JSX.Element}
 */
export const Slot = ({ slot, exclude, children, map }) => {
	if (slot && exclude) {
		throw new Error('Slot cannot be given a slot and exclude prop');
	}

	const match = React.Children.toArray(children).filter(
		exclude ? buildMatchFn(exclude, true) : buildMatchFn(slot)
	);

	return map ? match.map(map) : match;
};

/**
 * Iterate all children and map them with the slot they match
 *
 * @param {{slots: [Slot], map: ElementMapper, filter: ElementFilter}} props
 * @returns {JSX.Element}
 */
Slot.List = ({ slots, map = x => x, filter, children }) => {
	const matchers = slots.map(slot => ({
		slot,
		matches: buildMatchFn(slot),
	}));

	const filterFn = filter === true ? slot => !slot : filter;

	return React.Children.toArray(children).reduce((acc, child, index) => {
		const slot =
			matchers.find(m => m.matches(child, index))?.slot ?? undefined;

		return filterFn && !filterFn(slot, child, index)
			? acc
			: [...acc, map(slot, child, index)];
	}, []);
};

/**
 * Whether or not a child matches a slot
 *
 * @param {Slot} slot
 * @param {JSX.Element} child
 * @returns {boolean}
 */
Slot.matches = (slot, child) => buildMatchFn(slot)(child);

/**
 * Find the first child that matches the slot
 *
 * @param {Slot} slot
 * @param {React.Children} children
 * @returns {JSX.Element}
 */
Slot.find = (slot, children) =>
	React.Children.toArray(children).find(buildMatchFn(slot));

/**
 * Determine if a slot exists in children
 *
 * @param {Slot} slot
 * @param {React.Children} children
 * @returns {boolean}
 */
Slot.exists = (slot, children) =>
	React.Children.toArray(children).some(buildMatchFn(slot));

/**
 * Return the order of a given set of slots used in the children
 *
 * @param {Slot} slots
 * @param {React.Children} children
 * @returns {[Slot]}
 */
Slot.order = (slots, children) => {
	const childList = React.Children.toArray(children);
	const matchers = slots.map(slot => ({
		slot,
		matches: buildMatchFn(slot),
	}));

	return childList.map((child, ...args) => {
		const slot =
			matchers.find(m => m.matches(child, ...args))?.slot ?? undefined;

		return slot;
	});
};
