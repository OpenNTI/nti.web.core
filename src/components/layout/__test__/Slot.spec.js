/* eslint-env jest */
import TestRenderer from 'react-test-renderer';

import { Slot } from '../Slot';

const Aside = props => <div className="aside" {...props} />;

const Sample = props => (
	<section>
		<header>
			<Slot {...props} slot="header" />
		</header>
		<nav>
			<Slot {...props} slot="nav" />
		</nav>
		<aside>
			<Slot {...props} slot={Aside} />
		</aside>
		<p>
			<Slot {...props} slot="p" />
		</p>
		<footer>
			<Slot {...props} slot="footer" />
		</footer>
		<div className="unslotted">
			<Slot
				{...props}
				exclude={['header', 'nav', 'p', 'footer', Aside]}
			/>
		</div>
	</section>
);

const SlottedDiv = ({ slot, ...other }) => <div className={slot} {...other} />;

const SampleLayout = () => (
	<Sample>
		<SlottedDiv slot="header">Header</SlottedDiv>
		<SlottedDiv slot="nav">Nav</SlottedDiv>
		<Aside>Aside</Aside>
		<SlottedDiv slot="p">P</SlottedDiv>
		<SlottedDiv slot="footer">Footer</SlottedDiv>
		<span>Unslotted</span>
		<span>Unslotted</span>
		<span>Unslotted</span>
	</Sample>
);

describe('Slot Component', () => {
	describe('statics', () => {
		const named = <div slot="named" id="named" />;
		const cmp = <Aside id="cmp" />;
		const unslotted = <div id="unslotted" />;
		const children = [named, cmp, unslotted];

		describe('exists', () => {
			test('existing named', () =>
				expect(Slot.exists('named', children)).toBeTruthy());

			test('missing name', () =>
				expect(Slot.exists('non-existent', children)).toBeFalsy());

			test('existing cmp', () =>
				expect(Slot.exists(Aside, children)).toBeTruthy());

			test('missing cmp', () =>
				expect(Slot.exists(SampleLayout, children)).toBeFalsy());
		});

		describe('find', () => {
			test('existing named', () =>
				expect(Slot.find('named', children).props.id).toBe('named'));

			test('missing name', () =>
				expect(Slot.find('non-existent', children)).toBeUndefined());

			test('existing cmp', () =>
				expect(Slot.find(Aside, children).props.id).toBe('cmp'));

			test('missing cmp', () =>
				expect(Slot.find(SampleLayout, children)).toBeUndefined());
		});

		describe('order', () => {
			test('finds all children', () =>
				expect(
					Slot.order(['named', Aside, 'unused'], children).length
				).toBe(3));

			test('finds in children order', () => {
				const order = Slot.order([Aside, 'named'], children);

				expect(order).toEqual(['named', Aside, undefined]);
			});
		});
	});

	test('snapshot', () => {
		const testRender = TestRenderer.create(<SampleLayout />);
		expect(testRender.toJSON()).toMatchSnapshot();
	});

	test('renders children into the correct slots', () => {
		const testRender = TestRenderer.create(<SampleLayout />);

		const byClassName = (className, parent = testRender.root) =>
			parent.find(t => t.props.className === className);

		const byType = (type, parent = testRender.root) =>
			parent.findByType(type);

		const checkSlot = slot => {
			expect(() => byType(slot)).not.toThrow(); // should exist; e.g. <header>...</header>
			const el = byType(slot); // e.g. <header>...</header>

			expect(() => byClassName(slot, el)).not.toThrow(); // should exist; e.g. <div className="header">Header</div>
			const headerContent = byClassName(slot, el); // e.g. <div className="header">Header</div>;
			expect(headerContent.children[0]?.toLowerCase?.()).toEqual(slot); // text content
		};

		['header', 'nav', 'p', 'footer'].forEach(checkSlot);

		expect(() => byType(Aside)).not.toThrow();
		expect(() => byClassName('aside', byType(Aside))).not.toThrow();

		expect(byClassName('unslotted').findAllByType('span')).toHaveLength(3);
	});

	test('non-excluded unknown slots get included', () => {
		const testRender = TestRenderer.create(
			<Sample>
				<SlottedDiv slot="missing">Missing</SlottedDiv>
			</Sample>
		);

		const unslotted = testRender.root.find(
			t => t.props.className === 'unslotted'
		);

		expect(() => unslotted.findByType(SlottedDiv)).not.toThrow();
	});

	test('detects presence of children for the given slot', () => {
		const foo = <SlottedDiv key="foo" slot="foo" />;
		expect(Slot.exists('foo', [foo])).toBe(true);
		expect(Slot.exists('bar', [foo])).toBe(false);
	});
});
