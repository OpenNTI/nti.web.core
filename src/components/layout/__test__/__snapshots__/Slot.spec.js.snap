// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot Component snapshot 1`] = `
<section>
  <header>
    <div
      className="header"
    >
      Header
    </div>
  </header>
  <nav>
    <div
      className="nav"
    >
      Nav
    </div>
  </nav>
  <aside>
    <div
      className="aside"
    >
      Aside
    </div>
  </aside>
  <p>
    <div
      className="p"
    >
      P
    </div>
  </p>
  <footer>
    <div
      className="footer"
    >
      Footer
    </div>
  </footer>
  <div
    className="unslotted"
  >
    <span>
      Unslotted
    </span>
    <span>
      Unslotted
    </span>
    <span>
      Unslotted
    </span>
  </div>
</section>
`;
